Mods/SML/Source/SML/Private/Patching/BlueprintHookManager.cpp

```diff
@@ -35,102 +35,163 @@ void DebugDumpFunctionScriptCode(UFunction* Function, int32 HookOffset, const FS
+			}
+			break;
+		}
+	case EX_SwitchValue:
+		{
+			int32 CurrentJumpOffset = Expression->GetIntegerField(TEXT("OffsetToSwitchEnd"));
+			if (CurrentJumpOffset > HookOffset) {
+				// The switch end offset is just past the a single Word field holding the number of cases in the switch
+				int32 IndexOfCurrentJumpOffset = IndexAfterOpcode + sizeof(uint16);
+				FPlatformMemory::WriteUnaligned(&Script[IndexOfCurrentJumpOffset], (CodeSkipSizeType)(CurrentJumpOffset + UBlueprintHookManager::JumpBytesRequired));
+				IndexOfCurrentJumpOffset += sizeof(CodeSkipSizeType); // Move past the offset we just wrote
+				int32 SwitchExpressionSizeInBytes = (int32)Expression->GetObjectField(TEXT("Expression"))->GetNumberField(TEXT("ExpressionSizeInBytes"));
+				IndexOfCurrentJumpOffset += SwitchExpressionSizeInBytes; // Move past the switch expression
+				// Each case in the switch has an absolute offset reference to the next case, so we have to adjust each of them, as well
+				TArray<TSharedPtr<FJsonValue>> Cases = Expression->GetArrayField(TEXT("Cases"));
+				for (TSharedPtr<FJsonValue>& Case : Cases) {
+					TSharedPtr<FJsonObject> NextCase = Case->AsObject();
+					int32 CaseValueSizeInBytes = (int32)NextCase->GetObjectField(TEXT("CaseValue"))->GetNumberField(TEXT("ExpressionSizeInBytes"));
+					IndexOfCurrentJumpOffset += CaseValueSizeInBytes;
+					int32 OffsetToNextCase = (int32)NextCase->GetNumberField(TEXT("OffsetToNextCase"));
+					FPlatformMemory::WriteUnaligned(&Script[IndexOfCurrentJumpOffset], (CodeSkipSizeType)(OffsetToNextCase + UBlueprintHookManager::JumpBytesRequired));
+					IndexOfCurrentJumpOffset += sizeof(CodeSkipSizeType);
+					int32 CaseResultSizeInBytes = (int32)NextCase->GetObjectField(TEXT("CaseResult"))->GetNumberField(TEXT("ExpressionSizeInBytes"));
+					IndexOfCurrentJumpOffset += CaseResultSizeInBytes;
+				}
+			}
+			break;
+		}
+	case EX_AutoRtfmTransact:
+		{
+			int32 CurrentJumpOffset = Expression->GetIntegerField(TEXT("Offset"));
+			if (CurrentJumpOffset > HookOffset) {
+				// The offset is beyond a single TransactionId int
+				int32 IndexOfCurrentJumpOffset = IndexAfterOpcode + sizeof(int32);
+				FPlatformMemory::WriteUnaligned(&Script[IndexOfCurrentJumpOffset], (CodeSkipSizeType)(CurrentJumpOffset + UBlueprintHookManager::JumpBytesRequired));
+			}
+			break;
+		}
+	}
+
+	// Now we search all the children of this node for jump instructions that need to be updated
+	for (auto& Pair : Expression->Values) {
+		TSharedPtr<FJsonObject>* ExpressionValue;
+		if (Pair.Value->TryGetObject(ExpressionValue)) {
+			ModifyJumpTargetsForNewHookOffset(Script, *ExpressionValue, HookOffset);
+			continue;
+		}
+		TArray<TSharedPtr<FJsonValue>>* ArrayValue;
+		if (Pair.Value->TryGetArray(ArrayValue)) {
+			for (TSharedPtr<FJsonValue> Value : *ArrayValue) {
+				if (Value->TryGetObject(ExpressionValue)) {
+					ModifyJumpTargetsForNewHookOffset(Script, *ExpressionValue, HookOffset);
+				}
+			}
+		}
```

Comment:
Have you found any case in which this would do anything? I don't think jump instructions can appear in expressions (only in statements)

Reply:
I quickly looked over the Kismet compiler in the engine and it only emits `EX_JumpIfNot` from compiled statements (`KCST_` enumerators). I think the same applies to the other instructions with jump offsets. So this doesn't seem to be necessary.

Reply:
Yeah, my first pass didn't have the array handling here and was crashing.  Each case in a switch statement contains an OffsetToNextCase that is an absolute offset and needs to be updated, as well. Those are technically jump targets but perhaps the name of the function adds some confusion.  I changed it to ModifyOffsetsForNewHookOffset, which is hopefully a little more on the nose?

Reply:
OffsetToNextCase is an absolute offset? That's weird.

Reply:
Yeah, I was surprised, too.  Here's a snippet to show it and illustrate why recursing is necessary.  This LetBool's Expression is a switch and you can see each CaseValue's OffsetToNextCase is exactly the OpcodeIndex of the subsequent case.  Also, the OffsetToNextCase values are way too big to be relative offsets.

```
{
		"Opcode": 20,
		"OpcodeIndex": 258,
		"Inst": "LetBool",
		"Variable":
		{
			"Opcode": 72,
			"OpcodeIndex": 259,
			"Inst": "LocalOutVariable",
			"VariableType":
			{
				"PinCategory": "bool",
				"PinSubCategory": "bool"
			},
			"VariableName": "ReturnValue",
			"OpSizeInBytes": 9
		},
		"Expression":
		{
			"Opcode": 105,
			"OpcodeIndex": 268,
			"Inst": "SwitchValue",
			"Expression":
			{
				"Opcode": 0,
				"OpcodeIndex": 275,
				"Inst": "LocalVariable",
				"VariableType":
				{
					"PinCategory": "byte",
					"PinSubCategory": "byte",
					"PinSubCategoryObject": "/Script/FactoryGame.ERepresentationType"
				},
				"VariableName": "Temp_byte_Variable",
				"OpSizeInBytes": 9
			},
			"OffsetToSwitchEnd": 608,
			"Cases": [
				{
					"CaseValue":
					{
						"Opcode": 36,
						"OpcodeIndex": 284,
						"Inst": "ByteConst",
						"Value": 0,
						"OpSizeInBytes": 2
					},
					"OffsetToNextCase": 299,
					"CaseResult":
					{
						"Opcode": 0,
						"OpcodeIndex": 290,
						"Inst": "LocalVariable",
						"VariableType":
						{
							"PinCategory": "bool",
							"PinSubCategory": "bool"
						},
						"VariableName": "Temp_bool_Variable_20",
						"OpSizeInBytes": 9
					}
				},
				{
					"CaseValue":
					{
						"Opcode": 36,
						"OpcodeIndex": 299,
						"Inst": "ByteConst",
						"Value": 1,
						"OpSizeInBytes": 2
					},
					"OffsetToNextCase": 314,
					"CaseResult":
					{
						"Opcode": 0,
						"OpcodeIndex": 305,
						"Inst": "LocalVariable",
						"VariableType":
						{
							"PinCategory": "bool",
							"PinSubCategory": "bool"
						},
						"VariableName": "Temp_bool_Variable_19",
						"OpSizeInBytes": 9
					}
				},
				{
					"CaseValue":
					{
						"Opcode": 36,
						"OpcodeIndex": 314,
						"Inst": "ByteConst",
						"Value": 2,
						"OpSizeInBytes": 2
					},
					"OffsetToNextCase": 329,
					"CaseResult":
					{
						"Opcode": 0,
						"OpcodeIndex": 320,
						"Inst": "LocalVariable",
						"VariableType":
						{
							"PinCategory": "bool",
							"PinSubCategory": "bool"
						},
						"VariableName": "Temp_bool_Variable_18",
						"OpSizeInBytes": 9
					}
				},
			....
```

Reply:
Which asset is this snippet from? I'd like to inspect it myself

Reply:
That is from BPW_MapFilterCategories::CanBeSeenOnCompass and it's the first BeforeHook dump that is created with DEBUG_BLUEPRINT_HOOKING set to 1 (to ensure it was unmodified).

Reply:
Ack